pmkoalas/models/petrinet.py

```diff
@@ -0,0 +1,341 @@
+
+    def __repr__(self):
+        return 'Transition:' + str(self)
+
+
+SILENT_TRANSITION_DEFAULT_NAME='tau'
+def silentTransition(name=None,tid=None,weight=None):
+    tn = SILENT_TRANSITION_DEFAULT_NAME
+    if name:
+        tn = name
+    tw = 1
+    ttid = None
+    if tid:
+        ttid = tid
+    if weight:
+        tw = weight
+    return Transition(name=tn,weight=tw,tid=ttid,silent=True)
+
+
+class Arc:
+    def __init__(self,fromNode,toNode):
+        self._fromNode = fromNode
+        self._toNode = toNode
+
+    @property
+    def fromNode(self):
+        return self._fromNode
+
+    @property
+    def toNode(self):
+        return self._toNode
+
+    def __eq__(self,other):
+        if type(other) is type(self):
+            return self.fromNode == other.fromNode and \
+                    self.toNode  == other.toNode
+        return False
+
+    def __hash__(self):
+        return hash((self._fromNode,self._toNode))
+
+    def __str__(self):
+        return f'{self.fromNode} -> {self.toNode}' 
+
+    def __repr__(self):
+        return 'Arc:' + str(self)
+
+
+
+class PetriNetDOTFormatter:
+    def __init__(self,pn,font='SimSun'):
+        self._pn = pn
+        self._font = font
+        self._nodemap = {}
+        self._defaultHeight = 0.2
+
+    def tranDOT(self,tran,ti):
+        fstr = 'n{} [shape="box",margin="0, 0.1",label="{} {}",style="filled"];\n'
+        tl = tran.name if tran.name and tran.name != 'tau' else '&tau;'
+        fstr = f'n{str(ti)} [shape="box",margin="0, 0.1",'
+        fstr += f'label="{tran.weight}", style="filled",'
+        height = self._defaultHeight
+        fstr += f'height="{height}", width="{height}"'
+        fstr += '];\n'
+        return fstr
+
+    def placeDOT(self,place,pi):
+        fstr = '{} [shape="circle",label="{}"];\n'
+        return fstr.format('n' + str(pi),place.name)
+
+    def arcDOT(self,arc):
+        fromNode = self._nodemap[arc.fromNode]
+        toNode = self._nodemap[arc.toNode]
+        return f'n{fromNode}->n{toNode}\n'
+
+    def netToDOT(self):
+        dotstr = ""
+        dotstr += 'digraph G{\n'
+        dotstr += f'ranksep=".3"; fontsize="14"; remincross=true; margin="0.0,0.0"; fontname="{self._font}";rankdir="LR";charset=utf8;\n'
+        dotstr += 'edge [arrowsize="0.5"];\n'
+        dotstr += f'node [height="{self._defaultHeight}",width="{self._defaultHeight}",fontname="{self._font}"'
+        dotstr += ',fontsize="14"];\n'
+        dotstr += 'ratio=0.4;\n'
+        ni = 1
+        for pl in self._pn.places:
+            ni += 1
+            self._nodemap[pl] = ni
+            dotstr += self.placeDOT(pl,ni)
+        for tr in self._pn.transitions:
+            ni += 1
+            self._nodemap[tr] = ni
+            dotstr += self.tranDOT(tr,ni)
+        for ar in self._pn.arcs:
+            dotstr += self.arcDOT(ar)
+        dotstr += '}\n'
+        return dotstr
+
+
+def exportToDOT(net:LabelledPetriNet) -> str:
```

Comment:
should be a function of the structure.

Reply:
Hard disagree. Particular export formats need to be decoupled from the data structure. PNML is a standard for Petri nets which is the only reason it snuck under the wire. But the PNML functions could be moved out.

Reply:
This only introduces pointless import statements and requires the user to know more pointless information about the hard structure of the module. This is very counter intuitive to me.

Reply:
One thing I was considering was exporting some common functions across the packages in `__init__.py`. What are your thoughts on that idiom?

DOT comes with a load of external dependencies that really should stay out of the main data structure module. Same with pm4py, which I took out of the entire patch. Would like your thoughts about how to incorporate that, preferably without including pm4py as a dependency for koalas.